---
title: "Jointly define cell types from single cell multi-omics datasets"
author: "Jinzhuang Dou"
date: "10/14/2020"
output:
    html_document: default
    pdf_document: default
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  collapse = TRUE,  
  comment = "#>"
)
```

This guide will demonstrate the usage of bindSC to jointly define cell types by leveraging mulitple single-cell modalities. Here we define cell types that incorporate both gene expression and chromatin accessibility data. 

### Stage 1: Load the scRNA-seq and scATAC-seq data

**For convenience, we have prepared the pre-processed data which are ready to use.** 
User can refer to `A549_preprocess.ATAC.html` and `A549_preprocess.RNA.html` for the details of running the pre-processing workflow (It will take 10 ~ 20 mins). 

```{r, message = FALSE, warning=FALSE, results='hide'}
source("../../R/BiCCA.R")
source("../../R/bindSC.R")
source("../../R/ColorPalettes.R")
source("../../R/generics.R")
source("../../R/Heatmap.R")
source("../../R/preProcessing.R")


library(progress)
library(umap)
library(irlba)
library(Seurat)
library(dplyr)
library(ComplexHeatmap)
library(reshape2)
library(ggplot2)
library(reshape2)
library(Matrix)
library(dplyr)

# Change A549_rna.rds and A549_atac.rds to your own path

A549_RNA <- readRDS("../../../data/A549_rna.rds")
A549_ATAC <- readRDS("../../../data/A549_atac.rds")

summary(A549_RNA)
summary(A549_ATAC)

```

We then visualize each cell, colored by cell type, from two technologies. Left is from scRNA-seq and right is from scATAC-seq. For both technologies, cells from 0 h and 1/3 h can be well separated in 2D-UMAP. 

```{r, message = FALSE, warning=FALSE, results='hide',fig.height=3, fig.width=6}
library(ggpubr)
p1<-ggscatter(A549_RNA$RNA_meta, x = "UMAP_1", y = "UMAP_2",
   color = "cell_type", palette = c("darkseagreen4","lightpink","darkorchid1"), 
   repel = FALSE,size=0.5, alpha=0.5,legend.title = "", title="scRNA-seq", font.title=16) 

p2<-ggscatter(A549_ATAC$ATAC_meta, x = "UMAP_1", y = "UMAP_2",
   color = "group", palette = c("darkseagreen4","lightpink","darkorchid1"), 
   repel = FALSE,size=0.5, alpha=0.5,legend.title = "", title="scATAC-seq", font.title=16) 

p1<-p1+rremove("axis") + rremove("ticks") + rremove("xylab")+ rremove("axis.text") 
p2<-p2+rremove("axis") + rremove("ticks") + rremove("xylab")+ rremove("axis.text")

p<-ggarrange(p1, p2, nrow = 1, common.legend = TRUE, legend = "right")
p
```

### Stage 2: Preparion of bindSC input

We next perform bindSC on A549 data. bindSC requires three matrices as input: 

- `X` :  gene expression matrix from scRNA-seq data 
- `Y` :  chromatin accessibility matrix from scATAC-seq data 
- `Z0` : gene activity matrix from scATAC-seq data 

The gene activity matrix `Z0` is estimated by counting the total number of ATAC-seq reads within the gene body/promoter region of each gene in each cell. More details can be seen in `A549_preprocess.ATAC.html`

```{r, message = FALSE, warning=FALSE, results='hide'}

X <- A549_RNA$X
Y <- A549_ATAC$Y
Z0 <- A549_ATAC$Z0

```

Make sure `X` and `Z0` have matched gene features, `Y` and `Z0` have matched cell names. 

```{r, message = TRUE, warning=FALSE, results='markup'}

gene.overlap <- intersect(rownames(X), rownames(Z0))
cell.overlap <- intersect(colnames(Y), colnames(Z0))

X <- as.matrix(X[gene.overlap,])
Z0 <- as.matrix(Z0[gene.overlap, cell.overlap])
Y <- as.matrix(Y[,cell.overlap])
```

# Create bindSC object 

```{r , message = TRUE, warning=FALSE, results='markup'}

a549 <- createBindSCobj(X=X, Z0=Z0, Y=Y)


```

# Add meta data to available bindSC object. Parameter `name` specify the name of column added.  


```{r , message = TRUE, warning=FALSE, results='markup'}

treatmentTime <- c(A549_RNA$RNA_meta$treatment_time, A549_ATAC$ATAC_meta$group)
treatmentTime[treatmentTime=="A549_0h"] <- 0
treatmentTime[treatmentTime=="A549_1h"] <- 1
treatmentTime[treatmentTime=="A549_3h"] <- 3

type <- c(rep("scRNA-seq",nrow(A549_RNA$RNA_meta)), rep("scATAC-seq",nrow(A549_ATAC$ATAC_meta)))
a549 <- addMetaData(a549, treatmentTime,name="treatmentTime")
a549 <- addMetaData(a549, type,name="type")

# You can check the added column in meta data 

head(a549@meta)

```

### Stage 3: Align cell and features simultaneously using biCCA 

# Key parameters to run biCCA: 

- `k`: Number of canonical vectors to calculate [default `5`] 
- `alpha`: Couple coefficent [default 0.7]
- `block_size`: Sample/locus size for each block. This option is efficient when sample/locus size is very large (>30k) [default `5000`]
- `num.iteration`: Maximal number of iteration [default `100`]
- `tolerance`: Relative change ratio for imputed gene expression `Z` during iteration [default `0.01`]
- `temp.path`: Directory storing ouput results 
- `save`: Save the temporay files [default `FALSE`]


This process will take 25~30 mins with maximal memory usage being 15G. 

```{r , message = TRUE, warning=FALSE, results='markup'}

a549<- runBiCCA(a549,
                k=5, 
                alpha =0.7, 
                block.size =5000, 
                num.iteration = 100, 
                tolerance = 0.01, 
                temp.path = "tp" )
print("done")

```

# `runBiCCA` step added `biCCA` class to a549 object. There are 9 matrices inside: 

- `u`: contains the canonical correlation vectors for cells from first data (`K` cells by `k` factor) 
- `r`: contains the canonical correlation vectors for cells from second data (`L` cells by `k` factor)  
- `s`: contains the canonical correlation vectors for features from first data (`M` genes by `k` factor)  
- `v`: contains the canonical correlation vectors for features from second data (`N` loci by `k` factor) 
- `Z`: the imputed gene expression of second data 
- `delta`: relative change ratio for `Z` during iteration 
- `cost1`: sum of top k eigenvalues of SVD for cell alignment during iteration 
- `cost2`: sum of top k eigenvalues of SVD for feature alignment during iteration
- `umap`: 2D-UMAP of cells from two dataset



```{r, message = TRUE, warning=FALSE, results='markup'}

summary(a549@biCCA)
```

### Stage 4: Jointly clustering on two modalities 

Cell coordinates of two datasets in latent space are often sufficient to define joint clusters. 
There are two `resolution` parameters. `resolution1` sets the granularity of jointly clustering, with increased values leading to a greater number of clusters. `resolution2` set the granularity of pseudo cell contruction in each cluster.  

```{r, message = FALSE, warning=FALSE, results='hide',fig.height=5, fig.width=8}

a549 <- JointClustering(a549, resolution1 = 0.5, resolution2 = 2)

```

There will be two colummns `impuClst` and `pseudoCell` added on meta data 

```{r, message = FALSE, warning=FALSE, results='hide',fig.height=5, fig.width=8}
head(a549@meta)

```

### Stage 5: Jointly profiling of gene expression, chromatin accessiblity and TF activity in pseudo-cell level.  
This will take 5~10 mins.

```{r, message = FALSE, warning=FALSE, results='hide',fig.height=5, fig.width=8}

a549 <- PseudoCellProfiling(a549)

```

`a549` will have two matrics: `a549@pseudoCell_X` and `a549@pseudoCell_Y`. Each row denotes one gene/locus and each column denotes each pseudo cell. 

```{r, message = FALSE, warning=FALSE, results='hide',fig.height=5, fig.width=8}

a549@pseudoCell_X[1:5,1:5]

```

User can add new pseudo cell profile to `a549` object, for example, TF motif-based activity. We used `chromVAR` to calculate TF motif-based activity.   

```{r, message = FALSE, warning=FALSE, results='hide',fig.height=5, fig.width=8}
tf_chromVAR <- readRDS(file="../../data/a549.chromVAR.rds")
# make sure only pseudoCell from ATAC-seq data is loaded 
a549 <- AddTFPseudoCellProfiling(a549, as.matrix(tf_chromVAR), a549@meta$pseudoCell[a549@meta$type=="scATAC-seq"])

a549@pseudoCell_tf[1:5,1:5]
  
```

Heatmap shows joint profiles of gene expression, chromatin accessbilty and TF activity 


```{r, message = FALSE, warning=FALSE, results='hide',fig.height=7, fig.width=9}


# you can highlight genes/peaks/TF using `anno` option  (order is not required)
res <- PseudoCellPlot(a549, anno <- c("NR3C1", "MA0113.3_NR3C1"), top_gene = 50, top_peak = 50)
a549 <- res$obj
p <- res$plot
p

```


### Downstream analysis: comparsion among bindSC, Seurat, and Liger methods 

Three metrics are used to measure method performance since we have the cell correspondence as the gold standard.
**For convenience, we have prepared the pre-processed results for Seurat and Liger which are ready to use.** 

- `Silhoutte coefficient` : High value means cell-type architectures is well preserved
- `Alignment score` : High value means uniformity of mixing for two datasets in the latent space
- `Anchor accuracy` : High value means cell correspondence can be found in cell's k-neighbor size (k ranges from 5 o 500)


```{r, message = TRUE, warning=FALSE, results='markup'}
source("../A549/eval_plot.r")
cell_type <- a549@meta$treatmentTime
data_type <- a549@meta$type
result <- umap_plot(a549@biCCA$umap, cell_type, data_type)
eval <-  method_compare(result$plt_dt, "A549")
```

```{r plot1, echo=TRUE, fig.height=6, fig.width=8}
eval$coembedding

```

```{r plot2, echo=TRUE, fig.height=4, fig.width=8}
# Gray curve in anchor accuracy metric denotes results from random guess. 
eval$eval

```
